src/hotspot/share/runtime/arguments.cpp

```diff
@@ -1029,98 +1022,86 @@ const char* Arguments::handle_aliases_and_deprecation(const char* arg, bool warn
     }
-    JVMFlag* flag = JVMFlag::find_flag(real_name);
-    return set_bool_flag(flag, false, origin);
   }
-  if (sscanf(arg, "+%" XSTR(BUFLEN) NAME_RANGE "%c", name, &dummy) == 1) {
-    real_name = handle_aliases_and_deprecation(name, warn_if_deprecated);
-    if (real_name == NULL) {
-      return false;
+
+  const char* real_name = Arguments::handle_aliases_and_deprecation(name);
+  if (real_name == NULL) {
+    return NULL;
+  }
+  JVMFlag* flag = JVMFlag::find_flag(real_name);
+  return flag;
+}
+
+bool Arguments::parse_argument(const char* arg, JVMFlagOrigin origin) {
+  bool is_bool = false;
+  bool bool_val = false;
+  char c = *arg;
+  if (c == '+' || c == '-') {
+    is_bool = true;
+    bool_val = (c == '+');
+    arg++;
+  }
+
+  const char* name = arg;
+  while (true) {
+    c = *arg;
+    if (('a' <= c && c <= 'z' ) ||
+        ('A' <= c && c <= 'Z' ) ||
+        ('0' <= c && c <= '9' ) ||
+        ('_' == c)) {
+      ++arg;
+    } else {
+      break;
     }
-    JVMFlag* flag = JVMFlag::find_flag(real_name);
-    return set_bool_flag(flag, true, origin);
   }
 
-  char punct;
-  if (sscanf(arg, "%" XSTR(BUFLEN) NAME_RANGE "%c", name, &punct) == 2 && punct == '=') {
-    const char* value = strchr(arg, '=') + 1;
+  size_t name_len = size_t(arg - name);
+  JVMFlag* flag = find_jvm_flag(name, name_len);
+  if (flag == NULL) {
+    return false;
+  }
 
-    // this scanf pattern matches both strings (handled here) and numbers (handled later))
-    real_name = handle_aliases_and_deprecation(name, warn_if_deprecated);
-    if (real_name == NULL) {
+  if (is_bool) {
+    if (*arg != 0) {
+      // Error -- extra characters such as -XX:+BoolFlag=123
       return false;
     }
-    JVMFlag* flag = JVMFlag::find_flag(real_name);
-    if (flag != NULL && flag->is_ccstr()) {
+    return set_bool_flag(flag, bool_val, origin);
+  }
+
+  if (arg[0] == '=') {
+    const char* value = arg + 1;
+    if (flag->is_ccstr()) {
       if (flag->ccstr_accumulates()) {
         return append_to_string_flag(flag, value, origin);
       } else {
-        if (value[0] == '\0') {
-          value = NULL;
-        }
         return set_string_flag(flag, value, origin);
       }
+    } else if (flag->is_double()) {
+      return set_fp_numeric_flag(flag, value, origin);
     } else {
-      warn_if_deprecated = false; // if arg is deprecated, we've already done warning...
+      return set_numeric_flag(flag, value, origin);
     }
   }
 
-  if (sscanf(arg, "%" XSTR(BUFLEN) NAME_RANGE ":%c", name, &punct) == 2 && punct == '=') {
-    const char* value = strchr(arg, '=') + 1;
+  if (arg[0] == ':' && arg[1] == '=') {
     // -XX:Foo:=xxx will reset the string flag to the given value.
-    if (value[0] == '\0') {
-      value = NULL;
-    }
-    real_name = handle_aliases_and_deprecation(name, warn_if_deprecated);
-    if (real_name == NULL) {
-      return false;
-    }
-    JVMFlag* flag = JVMFlag::find_flag(real_name);
+    const char* value = arg + 2;
```

Comment:
I wasn't aware of this syntax and I can't see what the difference is between = and := in our code (this seems to be something from makefile variable setting logic!). ??